### ISSUE TITLE
l3draw: possibility to be not colour safe

### ISSUE THREAD

**Issue by Skillmon** [latex3/latex3#965]
When creating glyphs with the help of `l3draw` it would be nice if we could store them inside of a colour-unsafe box, so that they don't use the active colour when they were stored inside the box, but change the colour according to the surrounding text. Compare the following (in which I use a mix of TeX box syntax, LaTeX2e macros, and `l3draw` code):

```
\documentclass[]{article}

\usepackage{l3draw}

\ExplSyntaxOn
\newsavebox\rulebox
\newsavebox\drawbox
\setbox\drawbox=\hbox
  {
    \draw_begin:
      \draw_linewidth:n { 0.5pt }
      \draw_path_moveto:n { 0, 0 }
      \draw_path_lineto:n { 0, 1cm }
      \draw_path_lineto:n { 5mm, 1cm }
      \draw_path_use_clear:n { stroke }
    \draw_end:
  }
\setbox\rulebox=\hbox
  {
    \rule{0.5pt}{1cm}
    \rule[\dimexpr1cm-.5pt\relax]{5mm}{.5pt}
  }
\ExplSyntaxOff

\usepackage{xcolor}

\begin{document}
\usebox\rulebox\usebox\drawbox
\textcolor{red}{\usebox\rulebox\usebox\drawbox}
\end{document}
```

**Comment by josephwright** [comment]
Feels _a bit_ like something else I've not yet looked at from `pgf`: being able to store and reuse objects.  I guess I wonder if the right level here is 'store in a box', not least because LaTeX2e boxes also reset the current colour (you are using raw TeX primitives to get around that).

That said, it's easy enough to set up to avoid preserving colour in a drawing: I just have to add some customisation code to the start-of-scope line.

**Comment by Skillmon** [comment]
I know what I did, and why. Note that reusing parts of the image isn't what I'm asking about here. This is asked because it is: a) easy to set up from a user's point of view (and you're indicating it is easy to implement as well); and b) *very* performant. The workaround is simple: just don't box things up but insert the draw code each time, it'll pick up the colour itself. But that results in the picture being redrawn each and every time for no reason.

My current approach in my use case (which uses PGF currently, but the same holds true there) is to store the size and colour in a macro and check that against the current settings to determine whether I need to redraw the symbols or not. Gives not so bad performance overall, but of course it could be better.

**Comment by josephwright** [comment]
> I know what I did, and why. 

I'm thinking about how this request fits into the wider `expl3` environment. At present `\hbox_set:Nn` etc. don't force the current colour to apply, but that's partly as the matching `\vbox_...` situation is a lot more tricky. So it's at least possible that there might be no sensible way to use an `l3draw` feature here if it just works a 'raw' box setting. (The aim of course is that longer-term `expl3` should be entirely self-contained, which means no `\setbox\foo` outside of the bootstrapping of `l3box`.) At the LaTeX document level, `\savebox` already has that protection, so wouldn't work [I know you know that :)].

**Comment by josephwright** [comment]
>Note that reusing parts of the image isn't what I'm asking about here 

True, but other than reusing drawing elements, I can't think of where a 'no colour' option would show up: the colour otherwise matches the surroundings anyway.

**Comment by u-fischer** [comment]
> True, but other than reusing drawing elements, I can't think of where a 'no colour' option would show up: the colour otherwise matches the surroundings anyway.

Well it depends if you mean the surrounding when saving or when using.  

Currently, if `color` is loaded the boxes store the color when saving and so ignore the use-environment. Personally I think it would be useful sometimes to have the option to avoid this and get a box that adapts to the use-environment as you get if you use l3color instead:

~~~~
\documentclass{article}
%\usepackage{color}
\newsavebox\test
\begin{document}
\savebox\test{something}
\usebox\test \quad 
\ExplSyntaxOn
\color_select:n{red}
\ExplSyntaxOff
\usebox\test
\end{document}
~~~~

**Comment by josephwright** [comment]
> > True, but other than reusing drawing elements, I can't think of where a 'no colour' option would show up: the colour otherwise matches the surroundings anyway.
> 
> Well it depends if you mean the surrounding when saving or when using.
> 
> Currently, if `color` is loaded the boxes store the color when saving and so ignore the use-environment. Personally I think it would be useful sometimes to have the option to avoid this and get a box that adapts to the use-environment as you get if you use l3color instead:
> 
> ```
> \documentclass{article}
> %\usepackage{color}
> \newsavebox\test
> \begin{document}
> \savebox\test{something}
> \usebox\test \quad 
> \ExplSyntaxOn
> \color_select:n{red}
> \ExplSyntaxOff
> \usebox\test
> \end{document}
> ```

As I said, currently `expl3` boxes don't have `\colour_ensure_current:` so they are not 'completely safe', in contrast to coffins. I guess one issue is that there are two entirely opposing outcomes here, both with some logic behind them!

**Comment by davidcarlisle** [comment]
On Tue, 13 Jul 2021 at 16:54, Ulrike Fischer ***@***.***>
wrote:

> True, but other than reusing drawing elements, I can't think of where a
> 'no colour' option would show up: the colour otherwise matches the
> surroundings anyway.
>
> Well it depends if you mean the surrounding when saving or when using.
>
> Currently, if color is loaded the boxes store the color when saving and
> so ignore the use-environment. Personally I think it would be useful
> sometimes to have the option to avoid this and get a box that adapts to the
> use-environment as you get if you use l3color instead:
>
> \documentclass{article}
> %\usepackage{color}
> \newsavebox\test
> \begin{document}
> \savebox\test{something}
> \usebox\test \quad
> \ExplSyntaxOn
> \color_select:n{red}
> \ExplSyntaxOff
> \usebox\test
> \end{document}
>
> —
>
> in the original color design this was a deliberate design decision to
mimic font choices. In cases where you want something to pick up the
current colour you might also want them to pick up the current font, and
that's really not possible, the answer for fonts is not to use a box.

Of course colour seems different as at the implementation level (unless you
use lua attribute based color) the default would be to inherit the color at
point of use, but at the user level making color act like fonts isn't
entirely a bad thing.

For coloured fonts or luacolor style lua attribute color, it's more natural
to make it work like fonts and freeze the box (although no doubt it would
be possible to traverse the nodes in the box and re-color if you really
needed that)

**Comment by josephwright** [comment]
Ah, yes, Lua color ...

**Comment by car222222** [comment]
Do not lose sight of the original required use-case: 
When creating glyphs with the help of l3draw . . . 

It could be that 'a glyph' does (or does not) have an inbuilt  color: who knows?  

If there is a need to support such 'color-unspecified 
l3draw-based boxed-glyphs' then a new type of l3draw save box maybe needed.  But this is perhaps part of l3glyphs than l3draw?

**Comment by josephwright** [comment]
I suspect this is likely to be useful, as I've already added the idea of setting a default for line width. I wonder if the naming might need to be adjusted.

**Comment by josephwright** [comment]
I'm still wondering about this, in particular the LuaTeX attribute colour aspect. The latter makes any 'just let the general colour apply' approach a problem. Would welcome thoughts.
